commonmark/src/Commonmark/Types.hs

```diff
@@ -126,19 +126,30 @@ prettyRange (SourceRange xs) = go "" xs
     go _ [] = ""
     go curname ((p1,p2):rest)
       = (if sourceName p1 /= curname
-            then sourceName p1 ++ "@"
+            then escapeSourceName (sourceName p1) ++ "@"
             else "") ++
          show (sourceLine p1) ++ ":" ++
          show (sourceColumn p1) ++ "-" ++
          (if sourceName p2 /= sourceName p1
-             then sourceName p2 ++ "@"
+             then escapeSourceName (sourceName p2) ++ "@"
              else "") ++
          show (sourceLine p2) ++
          ":" ++ show (sourceColumn p2) ++
          if null rest
             then ""
             else ";" ++ go (sourceName p2) rest
 
+-- if the source name contains special characters it can lead to ambiguity when
+-- a filename exactly matches a fragment of syntax of the range
+escapeSourceName :: String -> String
+escapeSourceName = concatMap escapeChar
+  where
+    escapeChar '-' = "%-"
+    escapeChar '%' = "%%"
+    escapeChar ':' = "%:"
+    escapeChar ';' = "%;"
+    escapeChar x = [x]
+
```

Comment:
I was actally thinking of percent encoding like in URIs.
But it could be just for the characters you mention.  I think that's cleaner, because it's a recognizable format and people will know what's going on.
https://en.wikipedia.org/wiki/Percent-encoding